torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -137,6 +150,64 @@ Reducer::Reducer(
   }
 }
 
+void Reducer::mark_variable_ready_dense(
+    size_t replica_index,
+    size_t variable_index) {
+  const auto& bucket_index = variable_locators_[variable_index];
+  auto& bucket = buckets_[bucket_index.bucket_index];
+  auto& replica = bucket.replicas[replica_index];
+  auto& variable = replica.variables[bucket_index.intra_bucket_index];
+  const auto offset = replica.offsets[bucket_index.intra_bucket_index];
+  const auto length = replica.lengths[bucket_index.intra_bucket_index];
+
+  // Copy contents of gradient tensor to bucket tensor.
+  // If the gradient is not set, we assume it wasn't computed
+  // as part of the current backwards pass, and zero the part
+  // of the bucket it would otherwise hold.
+  auto bucket_view = replica.contents.narrow(0, offset, length);
+  auto& grad = variable.grad();
+  if (grad.defined()) {
+    // Ensure that the gradient type matches the bucket type.
+    AT_ASSERTM(
+        grad.type() == bucket_view.type(),
+        "Expected ",
+        bucket_view.type(),
+        ", got ",
+        grad.type());
+    // Assert that the grad tensor and the bucket don't share storage.
+    // If they did, we could avoid the copy altogether.
+    // The reason for not doing this is that existing code calls
+    // `detach_` from `zero_grad`, which is incompatible with views.
+    AT_ASSERT(!grad.is_alias_of(bucket_view));
+    AT_ASSERT(grad.device() == bucket_view.device());
+    AT_ASSERT(grad.numel() == bucket_view.numel());
+    bucket_view.copy_(grad.view({-1}), /* non_blocking */ true);
+  } else {
+    bucket_view.zero_();
+  }
+}
+
+void Reducer::mark_variable_ready_sparse(
+    size_t replica_index,
+    size_t variable_index) {
+  const auto& bucket_index = variable_locators_[variable_index];
+  auto& bucket = buckets_[bucket_index.bucket_index];
+  auto& replica = bucket.replicas[replica_index];
+  auto& variable = replica.variables[bucket_index.intra_bucket_index];
+  auto& grad = variable.grad();
+  AT_ASSERTM(
+      grad.options().layout() == c10::kSparse,
+      "Expected variable to have sparse gradient.");
+  AT_ASSERTM(grad.defined(), "Expected sparse gradient to be defined.");
```

Comment:
shall we do this for dense as well? Or is there any reason for the inconsistency between dense and sparse?

Reply:
Good point. Added the layout assert.